### PR TITLE
IGNITE-14209 Data rebalance on partition replicas' number changes

### PR DESCRIPTION
**Pull request by kgusakov**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/causality/VersionedValue.java [hunk 1]
```diff
@@ -594,7 +594,7 @@ private void trimToSize(long causalityToken) {
      * @param causalityToken Causality token.
      */
     private static void checkToken(long actualToken, long causalityToken) {
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/internal/causality/VersionedValue.java:596]: Not related to proposed changes but...
Both actualToken and causalityToken are causality tokens, I'd rather use "appliedToken" and "candidateToken" or similar. If you don't mind, lets fix it here. Besides that checkToken is rather assertToken, up to you though.

File: modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java [hunk 2]
```diff
@@ -143,6 +153,21 @@ public interface RaftGroupService {
...
+     * Changes peers of the replication group.
+     *
+     * <p>Asynchronous variant of the previous method.
+     * When the future completed, it just means, that changePeers process successfully started.
+     *
+     * <p>This operation is executed on a group leader.
+     *
+     * @param peers Peers.
+     * @param term Current known leader term.
+     *             If real raft group term will be different - changePeers will be skipped.
```
Comment (sanpwc) [inline modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/RaftGroupService.java:166]: I'd rather add few words about `org.apache.ignite.internal.raft.server.RaftGroupEventsListener#onNewPeersConfigurationApplied`, WDYT?

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java [hunk 3]
```diff
@@ -212,30 +210,71 @@ private CompletableFuture<RaftGroupService> prepareRaftGroupInternal(String grou
...
-     * @param lsnrSupplier Raft group listener supplier.
+     * @param grpId                   Raft group id.
+     * @param nodes                   Full set of raft group nodes.
+     * @param deltaNodes              New raft group nodes.
+     * @param lsnrSupplier            Raft group listener supplier.
+     * @param raftGrpEvtsLsnrSupplier Raft group events listener supplier.
      * @return Future representing pending completion of the operation.
      * @throws NodeStoppingException If node stopping intention was detected.
      */
     @Experimental
```
Comment (sanpwc) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:264]: It seems to be non experimental any more.

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java [hunk 4]
```diff
@@ -212,30 +210,71 @@ private CompletableFuture<RaftGroupService> prepareRaftGroupInternal(String grou
...
+     * If {@code deltaNodes} contains the current node, then raft group starts on the current node.
+     *
+     * @param grpId                   Raft group id.
+     * @param nodes                   Full set of raft group nodes.
+     * @param deltaNodes              New raft group nodes.
+     * @param lsnrSupplier            Raft group listener supplier.
+     * @param raftGrpEvtsLsnrSupplier Raft group events listener supplier.
+     * @throws NodeStoppingException If node stopping intention was detected.
+     */
+    public void startRaftGroupNode(String grpId, Collection<ClusterNode> nodes,
```
Comment (sanpwc) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/Loza.java:223]: Let's use one on a line approach here?

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupEventsListener.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import org.apache.ignite.raft.jraft.entity.PeerId;
+
+/**
+ * Listener for group membership and other events.
+ */
+public interface RaftGroupEventsListener {
+    /**
+     * Invoked, when new leader is elected (if it is the first leader of group ever - will be invoked too).
+     */
+    void onLeaderElected(long term);
```
Comment (sanpwc) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupEventsListener.java:31]: param javadoc?

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupEventsListener.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+     * Invoked on the leader, when new peers' configuration applied to raft group.
+     *
+     * @param peers list of peers, which was applied by raft group membership configuration.
+     */
+    void onNewPeersConfigurationApplied(List<PeerId> peers);
+
+    /**
+     * Invoked on the leader, when membership reconfiguration was failed, because of {@link Status}.
+     *
+     * @param status with description of failure.
```
Comment (sanpwc) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/server/RaftGroupEventsListener.java:43]: full params javadoc

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/Node.java [hunk 7]
```diff
@@ -189,6 +189,17 @@ public interface Node extends Lifecycle<NodeOptions>, Describer {
      */
     void changePeers(final Configuration newPeers, final Closure done);
 
+    /**
+     * Asynchronously change the configuration of the raft group to |newPeers|. If done closure was completed with {@link Status#OK()},
```
Comment (sanpwc) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/Node.java:193]: change**s**?

File: modules/table/tech-notes/rebalance.md [hunk 8]
```diff
@@ -96,6 +96,7 @@ metastoreInvoke: \\ atomic
         partition.assignments.pending = empty
     else:
         partition.assignments.pending = partition.assignments.planned
+        remove(partition.assignments.planned)
```
Comment (sanpwc) [inline modules/table/tech-notes/rebalance.md:99]: Hmm, is that really all changes we should introduce in order to match design and current actual implementation?

File: modules/core/src/main/java/org/apache/ignite/internal/causality/VersionedValue.java (hunk 1 above)
Comment (kgusakov) [reply modules/core/src/main/java/org/apache/ignite/internal/causality/VersionedValue.java:596]: WDYT about (actualToken, candidateToken)?

**Pull request closed without merging**